conn.go

```diff
@@ -32,18 +32,138 @@ import (
+// reused across multiple netlink operations, instead of opening and closing the
+// underlying netlink connection only for the duration of a single netlink
+// operation.
+func AsLasting() ConnOption {
+	return func(cc *Conn) {
+		// We cannot create the underlying connection yet, as we are called
+		// anywhere in the option processing chain and there might be later
+		// options still modifying connection behavior.
+		cc.lasting = true
+	}
+}
+
+// WithNetNSFd sets the network namespace to create a new netlink connection to:
+// the fd must reference a network namespace.
+func WithNetNSFd(fd int) ConnOption {
+	return func(cc *Conn) {
+		cc.NetNS = fd
+	}
+}
+
+// WithTestDial sets the specified nltest.Func when creating a new netlink
+// connection.
+func WithTestDial(f nltest.Func) ConnOption {
+	return func(cc *Conn) {
+		cc.TestDial = f
+	}
+}
+
+// netlinkCloser is returned by netlinkConn(UnderLock) and must be called after
+// being done with the returned netlink connection in order to properly close
+// this connection, if necessary.
+type netlinkCloser func() error
+
+// netlinkConn returns a netlink connection together with a netlinkCloser that
+// later must be called by the caller when it doesn't need the returned netlink
+// connection anymore. The netlinkCloser will close the netlink connection when
+// necessary. If New has been told to create a lasting connection, then this
+// lasting netlink connection will be returned, otherwise a new "transient"
+// netlink connection will be opened and returned instead. netlinkConn must not
+// be called while the Conn.mu lock is currently helt (this will cause a
+// deadlock). Use netlinkConnUnderLock instead in such situations.
+func (cc *Conn) netlinkConn() (*netlink.Conn, netlinkCloser, error) {
+	cc.mu.Lock()
+	defer cc.mu.Unlock()
+	return cc.netlinkConnUnderLock()
+}
+
+// netlinkConnUnderLock works like netlinkConn but must be called while holding
+// the Conn.mu lock.
+func (cc *Conn) netlinkConnUnderLock() (*netlink.Conn, netlinkCloser, error) {
+	if cc.nlconn != nil {
+		return cc.nlconn, func() error { return nil }, nil
+	}
+	nlconn, err := cc.dialNetlink()
+	if err != nil {
+		return nil, nil, err
+	}
+	return nlconn, func() error { return nlconn.Close() }, nil
+}
+
+// Close closes the lasting netlink connection that has been opened using
+// AsLasting option when creating this connection. If either no lasting netlink
+// connection has been opened or the (not-so) lasting connection is already in
+// the process of closing or has been closed, Close will immediately return
+// without any error.
+//
+// Close will terminate all pending netlink operations using the lasting
+// connection.
+//
+// After closing a lasting connection, the connection will revert to using
+// on-demand transient netlink connections when calling further netlink
+// operations (such as GetTables).
+func (cc *Conn) Close() error {
```

Comment:
Given that `Close` only has an effect on the lasting connection, can we rename it to `CloseLasting`?

Reply:
While this is an unusual name I see the reason in this particular case, because we have the unusual situation of both on-demand as well as lasting connections. I've refactored the name.